packages/ledger/src/index.ts

```diff
@@ -0,0 +1,270 @@
+
+/**
+ * Returns the correct ledger transport based on browser compatibility for webUSB.
+ * @returns
+ */
+const getTransport = async () =>
+  ((await supportsWebUSB())
+    ? (await import('@ledgerhq/hw-transport-webusb')).default
+    : (await import('@ledgerhq/hw-transport-u2f')).default
+  ).create()
+
+interface LedgerAccount {
+  publicKey: string
+  derivationPath: string
+  chainCode: string
+}
+
+const getBalance = async (
+  address: string,
+  rpcUrl: string,
+  block: string | number = 'latest'
+): Promise<string> => {
+  const { providers } = await import('ethers')
+  const provider = new providers.JsonRpcProvider(rpcUrl)
+  return (await provider.getBalance(address, block)).toHexString()
+}
+
+const getAddress = async (
+  { publicKey, chainCode, derivationPath }: LedgerAccount,
+  asset: Asset,
+  { rpcUrl }: Chain,
+  index: number
+): Promise<Account> => {
+  const { BIP32Factory } = await import('bip32')
+  const ecc = await import('tiny-secp256k1')
+  const { Buffer } = await import('buffer')
+  const { publicToAddress, toChecksumAddress } = await import('ethereumjs-util')
+
+  const node: BIP32Interface = BIP32Factory(ecc).fromPublicKey(
+    Buffer.from(publicKey, 'hex'),
+    Buffer.from(chainCode, 'hex')
+  )
+
+  const child: BIP32Interface = node.derive(index)
+
+  const address = toChecksumAddress(
+    `0x${publicToAddress(child.publicKey, true).toString('hex')}`
+  )
+
+  console.log({
+    address,
+    index,
+    derivationPath,
+    fullPath: `${derivationPath}/${index}`
+  })
+
+  return {
+    derivationPath,
+    address,
+    balance: {
+      asset: asset.label,
+      value: await getBalance(address, rpcUrl)
+    }
+  }
+}
+
+const getAddresses = (
+  account: LedgerAccount,
+  asset: Asset,
+  currentChain: Chain,
+  offset: number = 0,
+  limit: number = 15
+): Promise<Account[]> =>
+  limit - offset <= 0
+    ? (() => {
+        throw new Error('Offset must be less than limit')
+      })()
+    : Promise.all(
+        Array.from({ length: limit - offset }, (_, index) =>
+          getAddress(account, asset, currentChain, index)
+        )
+      )
+
+function ledger(): WalletInit {
+  return () => {
+    let accounts: Account[] | undefined
+    return {
+      label: 'Ledger',
+      getIcon: async () => (await import('./icon')).default,
+      getInterface: async ({ EventEmitter, chains }) => {
+        const Eth = (await import('@ledgerhq/hw-app-eth')).default
+        const { TransactionFactory: Transaction, Capability } = await import(
+          '@ethereumjs/tx'
+        )
+        const { default: Common, Hardfork } = await import('@ethereumjs/common')
+        const { rlp } = (await import('ethereumjs-util')).default
+
+        const transport: Transport = await getTransport()
+        const eth = new Eth(transport)
+        const eventEmitter = new EventEmitter()
+
+        let currentChain: Chain | undefined
+        const scanAccounts = async ({
+          derivationPath,
+          chainId,
```

Comment:
@aaronbarnardsound As far as I can tell this is the only way we know what chainId the user has selected correct?